src/app/qgisapp.cpp

```diff
@@ -1941,6 +1954,143 @@ QgsMessageBar *QgisApp::visibleMessageBar()
   }
 }
 
+QList<QgsVectorLayerRef> QgisApp::findBrokenWidgetDependencies( QgsVectorLayer *vl )
+{
+  QList<QgsVectorLayerRef> brokenDependencies;
+  // Check for missing layer widget dependencies
+  for ( int i = 0; i < vl->fields().count(); i++ )
+  {
+    std::unique_ptr<QgsEditorWidgetWrapper> ww;
+    ww.reset( QgsGui::editorWidgetRegistry()->create( vl, i, nullptr, nullptr ) );
+    // ww should never be null in real life, but it is in QgisApp tests because
+    // QgsEditorWidgetRegistry widget factories is empty
+    if ( ww )
+    {
+      const auto constDependencies { ww->layerDependencies() };
+      for ( const QgsVectorLayerRef &dependency : constDependencies )
+      {
+        const QgsVectorLayer *depVl { QgsVectorLayerRef( dependency ).resolveWeakly( QgsProject::instance() ) };
+        if ( ! depVl || ! depVl->isValid() )
+        {
+          brokenDependencies.append( dependency );
+        }
+      }
+    }
+  }
+  return brokenDependencies;
+}
+
+void QgisApp::checkVectorLayerDependencies( QgsVectorLayer *vl )
+{
+  if ( vl && vl->isValid() )
+  {
+    const auto constDependencies { findBrokenWidgetDependencies( vl ) };
+    for ( const QgsVectorLayerRef &dependency : constDependencies )
+    {
+      const QgsVectorLayer *depVl = nullptr;
+      // The default resolution logic does not recognize a positive match if the datasource is
```

Comment:
This logic should belong in _LayerRef

Reply:
Yeah, I though about it, not sure: do you think at adding another method that `resolveAggresively` ?

But the main reason I wanted it in app and not in core was that I wanted to communicate with the user, what would you suggest? A signal connected to app ?

Reply:
I just mean the first part of this branch -- the part where it scans through project layers for ones which match the name and provider

Reply:
Sorry, I don't follow you: `resolveWeakly` does that (but it matches the public source exactly), the only option I see (and that's the one I proposed in my previous comment) is to have a `resolveAggressively` (or whatever you want to name it) that also searches for matching layers by provider AND name. But that won't be sufficient to solve my problem here, because _LayerRef is designed to find layers already defined in the current project while what I'm doing here is searching for layers in the provider that do not exist in the project.